src/ripple/rpc/handlers/WalletPropose.cpp

```diff
@@ -35,14 +38,58 @@ Json::Value walletPropose (Json::Value const& params)
     RippleAddress   naSeed;
     RippleAddress   naAccount;
 
-    if (!params.isMember (jss::passphrase))
-        naSeed.setSeedRandom ();
+    KeyType type = KeyType::secp256k1;
 
-    else if (!naSeed.setSeedGeneric (params[jss::passphrase].asString ()))
+    bool const has_key_type   = params.isMember (jss::key_type);
+    bool const has_passphrase = params.isMember (jss::passphrase);
+
+    if (has_key_type)
+    {
+        // `key_type` must be valid if present.
+
+        type = keyTypeFromString (params[jss::key_type].asString());
+
+        if (isInvalid (type))
+        {
+            return rpcError (rpcBAD_SEED);
+        }
+
+        naSeed = getSeedFromRPC (params);
+    }
+    else if (has_passphrase)
+    {
+        naSeed.setSeedGeneric (params[jss::passphrase].asString());
+    }
+    else
+    {
+        naSeed.setSeedRandom();
+    }
+
+    if (!naSeed.isSet())
+    {
         return rpcError(rpcBAD_SEED);
+    }
+
+    if (type == KeyType::secp256k1)
+    {
+        RippleAddress naGenerator = RippleAddress::createGeneratorPublic (naSeed);
+        naAccount.setAccountPublic (naGenerator, 0);
+    }
+    else if (type == KeyType::ed25519)
+    {
+        uint256 secretkey = keyFromSeed (naSeed.getSeed());
+
+        Blob publickey (33);
+        publickey[0] = 0xED;
+        ed25519_publickey (secretkey.data(), &publickey[1]);
+        secretkey.zero();  // security erase
 
-    RippleAddress naGenerator = RippleAddress::createGeneratorPublic (naSeed);
-    naAccount.setAccountPublic (naGenerator, 0);
+        naAccount.setAccountPublic (publickey);
+    }
+    else
+    {
+        assert (!!! "not reached");
```

Comment:
With `assert(! "not reached");` I try to parse it as "not not reached", which is confusing.  Whereas the triple bang is obviously meant to have a connotation rather than be read literally.

We should probably just have a `NOT_REACHED()` macro.